### PR TITLE
Clarify authorize vs sign

### PR DESCRIPTION
**Pull request by liamdennehy**
This PR is primarily in response to https://github.com/w3c-dvcg/http-signatures/issues/13, but also provides some general clarifications and rewording of intent:

* Explicitly define Authorization and Signature schemes to differentiate them and propose distinct purposes
* Clarify some language that assumes the server is the verifier and client is the signer, to make the spec more portable and usable by both
* Highlight this spec is useful for both website and API, not just API
* Highlight the possibility of unknown parties to perform dentification via a trusted CA
* Highlight RSA's ability to provide non-repudiation over shared secrets and most transports
* Explain why message (header) may be altered in transit and why signature string is useful over signing the entire message
* Some language in Signature chapter seem to cater exclusively or dominantly to authorisation and not general-purpose digital signatures
* Remove bias towards client signing and server verifying
* Highlight benefits of PKI for non-repudation
* Generalise language on keyId (meaningful to server, not necessarily known to server)
* typos

### CODE REVIEW

File: index.xml [hunk 1]
```diff
@@ -113,38 +113,66 @@ are no techniques in this proposal that are novel beyond the previous art, it
...
+This specification presents two mechanisms with distinct purposes:
+    <list style="numbers">
+       <t>
+The "Signature" scheme which is intended primarily to allow a sender
+to assert the contents of the message sent are correct and have not been
+altered during transmission or storage in a way that alters the meaning expressed
+in the original message as signed. Any party reading the message (the verifier)
+may independently confirm the validity of the message signature. This scheme
+is agnostic to the client/server direction and can be used to verify the contents of
+either HTTP requests, HTTP reponses or both.
```
Comment (msporny) [inline index.xml]: "reponses or" => "responses, or both"

File: index.xml [hunk 2]
```diff
@@ -113,38 +113,66 @@ are no techniques in this proposal that are novel beyond the previous art, it
...
+       </t>
+       <t>
+The "Authorization" scheme which is intended primarily to allow a sender
+to assert an identity and request access to a resource or resources. This is
+typically used in authentication flows such as OAuth2, and as a consequence
```
Comment (msporny) [inline index.xml]: This second sentence is hard to understand. Why are we mentioning OAuth2 - this spec isn't used in OAuth2.

File: index.xml [hunk 3]
```diff
@@ -113,38 +113,66 @@ are no techniques in this proposal that are novel beyond the previous art, it
...
+`Authorization` header is normally generated (and the message signed) by the
+HTTP client and the message verified by an HTTP service.
```
Comment (msporny) [inline index.xml]: We need to be careful about mixing "client" with "service". IETF is typically client/server. I'm not sure this second sentence adds much, or it needs to be reworded to be more clear.

File: index.xml [hunk 4]
```diff
@@ -113,38 +113,66 @@ are no techniques in this proposal that are novel beyond the previous art, it
...
+       </t>
+     </list>
+  </t>
 
   <section anchor="intro-requests" title="Using Signatures in HTTP Requests">
    <t>
-It is common practice to protect sensitive website API functionality via
+It is common practice to protect sensitive website and API functionality via
```
Comment (msporny) [inline index.xml:143]: This presumes that HTTP Signatures may be used in website flows, which it is not intended for (and some folks might mistakenly believe that we're trying to compete with WebAuthn/FIDO flows, which would be entering into a political minefield).

File: index.xml [hunk 5]
```diff
@@ -113,38 +113,66 @@ are no techniques in this proposal that are novel beyond the previous art, it
...
    </t>
    <t>
-Digital signatures are widely used to provide authentication without
+Digital signatures are widely used to provide authentication and integrity
+assurances without
 the need for shared secrets. They also do not require a round-trip in
-order to authenticate the client. A server need only have a mapping between
-the key being used to sign the content and the authorized entity to
+order to authenticate the client, and allow the integrity of a message to
+be verified independently of the transport (e.g. TLS)
```
Comment (msporny) [inline index.xml]: Missing period?

File: index.xml [hunk 6]
```diff
@@ -113,38 +113,66 @@ are no techniques in this proposal that are novel beyond the previous art, it
...
+the key being used to sign the content and the authorized entity) to
 verify that a message was signed by that entity.
    </t>
    <t>
-This specification provides two mechanisms that can be used by a server
-to authenticate a client. The first is the 'Signature' HTTP Authentication
-Scheme, which may be used for interactive sessions. The second is the
-Signature HTTP Header, which is typically used by automated software
-agents.
+When optionally combined with asymmetric keys in certificates issued
```
Comment (msporny) [inline index.xml]: Strike "in certificates issued via PKI" ... while that's possible, it's not only limited to that... see the Decentralized Identifiers work, which is what sparked the creation of this specification: https://w3c-ccg.github.io/did-primer/

File: index.xml [hunk 7]
```diff
@@ -113,38 +113,66 @@ are no techniques in this proposal that are novel beyond the previous art, it
...
+via PKI, this specification can also enable authentication of a client and
+server with no prior knowledge of each other via a trusted CA issuing their
```
Comment (msporny) [inline index.xml]: strike "via a trusted CA issuing their respective certificates" ... for the same reason as above.

File: index.xml [hunk 8]
```diff
@@ -113,38 +113,66 @@ are no techniques in this proposal that are novel beyond the previous art, it
...
+respective certificates.
    </t>
   </section>
 
   <section anchor="intro-responses" title="Using Signatures in HTTP Responses">
    <t>
-For high security transactions, having an additional signature on the HTTP
-header allows a client to ensure that even if the transport channel has been
-compromised, that the content of the messages have not been compromised.
+For high security transactions, having a mechanism to sign HTTP responses
```
Comment (msporny) [inline index.xml]: This sentence is a bit too complex. Refactor, please.

File: index.xml [hunk 9]
```diff
@@ -232,6 +261,24 @@ MUST be ignored.
 
    <section anchor="canonicalization" title="Signature String Construction">
 
+   <t>
+A signed HTTP message needs to be tolerant of some
+alterations during transmission
+as it goes through gateways, proxies and other entities. As such, the raw
```
Comment (msporny) [inline index.xml]: "proxies and" -> "proxies, and" -- use oxford comma throughout spec.

File: index.xml [hunk 10]
```diff
@@ -232,6 +261,24 @@ MUST be ignored.
...
+HTTP message transmitted by the signer, and the HTTP message received
+by the verifier may have undergone some changes, typically in the form of
```
Comment (msporny) [inline index.xml]: Comma usage... the sentence ends up having too much punctuation... refactor into simpler to understand sentences.

File: index.xml [hunk 11]
```diff
@@ -232,6 +261,24 @@ MUST be ignored.
...
+added, removed or altered headers. These changes are not normally under the
+control of the sender, and often the verifier. While these changes should not
+alter the meaning expressed by the sender this protocol allows for any sensitive
```
Comment (msporny) [inline index.xml]: "sender this" => "sender, this"

File: index.xml [hunk 12]
```diff
@@ -525,9 +576,18 @@ content-length: 18
  <section anchor="sig" title="The 'Signature' HTTP Header">
 
   <t>
-The "signature" HTTP Header is based on the model that the sender must
-authenticate itself with a digital signature produced by either a private
-asymmetric key (e.g., RSA) or a shared symmetric key (e.g., HMAC).  The scheme
+The "Signature" HTTP Header allows the the sender to add a signature
```
Comment (msporny) [inline index.xml]: The resulting sentence, while accurate, forces a high cognitive load when trying to logically process the statements. I suggest it is split into multiple, simpler sentences.

**Review by msporny — CHANGES_REQUESTED** [review]
Overall, I agree with the content. I'm trying very hard to keep the spec as simple as possible - minimalist in feature set, most sentences only requiring a ~7th grade reading level, avoiding unnecessarily flowery language, short spec length, etc.

Some of the explanatory text introduced in this PR should be simplified to try to achieve some of the readability goals of the spec. Some of the goals are not always possible, but we should strive for simplicity, minimalism, and clarity.

File: index.xml (hunk 1 above)
Comment (liamdennehy) [reply index.xml]: Resolved 525485088aecd30c9bef727dc8bcd9345cdfaa3d

File: index.xml (hunk 3 above)
Comment (liamdennehy) [reply index.xml]: Resolved 525485088aecd30c9bef727dc8bcd9345cdfaa3d

File: index.xml (hunk 5 above)
Comment (liamdennehy) [reply index.xml]: Resolved 525485088aecd30c9bef727dc8bcd9345cdfaa3d

File: index.xml (hunk 9 above)
Comment (liamdennehy) [reply index.xml]: Resolved 525485088aecd30c9bef727dc8bcd9345cdfaa3d

File: index.xml (hunk 2 above)
Comment (liamdennehy) [reply index.xml]: Resolved 525485088aecd30c9bef727dc8bcd9345cdfaa3d

File: index.xml (hunk 4 above)
Comment (liamdennehy) [reply index.xml:143]: From a server viewpoint, websites and APIs are not distinct - some web clients might invoke API to populate content. Just as a browser might prompt a user for a certificate to complete Mutual TLS, there's no reason client code wouldn't use a client's certificate to sign a request inside a web session. Note this is not just for authentication, but also to produce a digitally signed request.

Especially when used in a POST, this has particular power (e.g. upload a copy of your passport which can be a signed transaction). I don't want to use language prohibiting these use cases. Part of the justification for this PR is to remove the bias towards the authentication case over being able to digitally sign a request for the explicit purpose of turning a HTTP message into a digitally signed document - something no other protocol can do today.

File: index.xml (hunk 8 above)
Comment (liamdennehy) [reply index.xml]: Wow, that is indeed a lot of content for two sentences. Thx.
New attempt in 59f4a584358b1cac05b471b07a3844c360702c57

File: index.xml (hunk 2 above)
Comment (liamdennehy) [reply index.xml]: OAuth 2.0 is however introduced earlier:
https://github.com/w3c-dvcg/http-signatures/blob/18a4cbdb350ac2f02826e5567995fefe52c095fb/index.xml#L98

File: index.xml (hunk 6 above)
Comment (liamdennehy) [reply index.xml]: "asymmetric keys linked to an identity"?

File: index.xml (hunk 11 above)
Comment (liamdennehy) [reply index.xml]: Refactored in 59f4a584358b1cac05b471b07a3844c360702c57

File: index.xml (hunk 10 above)
Comment (liamdennehy) [reply index.xml]: Refactored in 59f4a584358b1cac05b471b07a3844c360702c57

File: index.xml (hunk 12 above)
Comment (liamdennehy) [reply index.xml]: Refactored in ac26524156a4f14652de28d1da2b66d7685d01f5

File: index.xml (hunk 7 above)
Comment (liamdennehy) [reply index.xml]: New wording to generalise the use case beyond PKI in:
https://github.com/w3c-dvcg/http-signatures/blob/03163a4e70ce8d3d27c8880195301bbe5d1dd8f0/index.xml#L165

File: index.xml (hunk 6 above)
Comment (liamdennehy) [reply index.xml]: New wording to generalise the use case beyond PKI in:
https://github.com/w3c-dvcg/http-signatures/blob/03163a4e70ce8d3d27c8880195301bbe5d1dd8f0/index.xml#L165

File: index.xml [hunk 13]
```diff
@@ -113,41 +113,73 @@ are no techniques in this proposal that are novel beyond the previous art, it
...
    </t>
    <t>
-This specification provides two mechanisms that can be used by a server
-to authenticate a client. The first is the 'Signature' HTTP Authentication
-Scheme, which may be used for interactive sessions. The second is the
-Signature HTTP Header, which is typically used by automated software
-agents.
+When optionally combined with asymmetric keys associated with an
+identity, this specification can also enable authentication of a client and
+server with or without no prior knowledge of each other.
```
Comment (bblfish) [inline index.xml]: "with or without no prior knowledge" there is a double negation here, that is unecessary. I think one can drop the word "no"

File: index.xml [hunk 14]
```diff
@@ -232,6 +265,21 @@ MUST be ignored.
...
+alterations during transmission
+as it goes through gateways, proxies, and other entities. These changes are often
+of little consequence and very benign, but also often not visible to or
+detectable by either the sender or the recipient. Simply signing the entire
+message that was transmitted by the sender is therefore not feasible:
+Even very minor changes would result in a signature which cannot be verified.
+   </t>
+   <t>
+This specification allows the sender to select which headers are meaningful
+by including their names the `headers` signature parameter. The headers
```
Comment (bblfish) [inline index.xml]: missing "in". It should be "by including their names in the `headers` signature"

File: index.xml [hunk 15]
```diff
@@ -422,17 +473,18 @@ Note that the use of the `Digest` header field is per
...
    </t>
 
    <section anchor="auth-isa" title="Initiating Signature Authorization">
 
     <t>
-A server may notify a client when a protected resource could be accessed by
-authenticating itself to the server. To initiate this process, the server
-will request that the client authenticate itself via a 401 response code. The
+A server may notify a client when a resource is protected by requiring a signature,
+To initiate this process, the server
```
Comment (bblfish) [inline index.xml:520]: should that be a full stop above?

Comment (liamdennehy) [reply index.xml:520]: Resolved 54dc50addae904b8a42cd152456bf53fdf438256

File: index.xml (hunk 14 above)
Comment (liamdennehy) [reply index.xml]: Resolved 54dc50addae904b8a42cd152456bf53fdf438256

File: index.xml (hunk 13 above)
Comment (liamdennehy) [reply index.xml]: Resolved 54dc50addae904b8a42cd152456bf53fdf438256

**Comment by liamdennehy** [comment]
@bblfish thanks for the detailed review

**Comment by liamdennehy** [comment]
Rebased to remove conflicts with recent PR merge.

File: index.xml (hunk 4 above)
Comment (liamdennehy) [reply index.xml:143]: @msporny Is this resolved, or do you need further clarification/edits?

File: index.xml (hunk 6 above)
Comment (liamdennehy) [reply index.xml]: I hope this is good and can be resolved?

File: index.xml (hunk 7 above)
Comment (liamdennehy) [reply index.xml]: I hope this is good and can be resolved?

File: index.xml (hunk 4 above)
Comment (liamdennehy) [reply index.xml:143]: bump...

**Review by msporny — APPROVED** [review]
Merging. Sorry this took so long @liamdennehy!

File: index.xml [hunk 16]
```diff
@@ -242,6 +275,21 @@ MUST be ignored.
 
    <section anchor="canonicalization" title="Signature String Construction">
 
+   <t>
+A signed HTTP message needs to be tolerant of some trivial
+alterations during transmission
+as it goes through gateways, proxies, and other entities. These changes are often
```
Comment (ioggstream) [inline index.xml:281]: @liamdennehy nitpicking: "gateways, proxies and other entities" rfc 7230 probably uses the term Intermediaries https://tools.ietf.org/html/rfc7230#section-2.3

Comment (liamdennehy) [reply index.xml:281]: I pick nits for a living, more than happy to be on the receiving end. Correcting in a general quality and fixup PR.

**Pull request merged**